### PR TITLE
ec2_instance test cleanup

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY

Attempts to stabilise the ec2_instance tests (see also #58650)

- Cleanup docs
- Migrate tests to use module_defaults
- Make sure we clean up resources when we're finished with them
- Pause after creating roles to ensure creation actually completed

**CC**: @s-hertel 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/amazon/ec2_instance.py
test/sanity/ignore.txt

##### ADDITIONAL INFORMATION

Found one 'interesting' quirk - 'wait' only waits for the completion of actions initiated by the specific call.
If you pass a filter or multiple instance_ids and they're already transitioning to the new state, then 'wait' will return straight away rather than waiting for the instances to reach the desired state.  This is problematic when dealing with mass-removals where you need to wait for dependencies to be 'gone' before deleting them (eg waiting on instance termination before deleting ENIs/volumes)

### CODE REVIEW

**Comment by tremble** [comment]
I believe the code's good at this point, so reviews would be appreciated.

Over the course of the next couple of hours I'm going to try to trigger additional test runs in the Shippable environment by rebasing, cleaning up unused/duplicate files and tweaking the documentation a little bit.  However, I believe I've tracked down the cause of the instability...

**Comment by tremble** [comment]
@s-hertel I'll push more rebases through out the day, but I think the tests are now consistently passing.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 1]
```diff
@@ -75,65 +85,123 @@
...
-      - If specifications for a single network are given, accepted keys are assign_public_ip (bool),
-        private_ip_address (str), ipv6_addresses (list), source_dest_check (bool), description (str),
-        delete_on_termination (bool), device_index (int), groups (list of security group IDs),
-        private_ip_addresses (list), subnet_id (str).
-      - I(network.interfaces) should be a list of ENI IDs (strings) or a list of objects containing the key I(id).
-      - Use the ec2_eni to create ENIs with special settings.
+      - Use the ec2_eni module to create ENIs with special settings.
+    type: dict
+    suboptions:
+      interfaces:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py:131]: It seems like a few things are now missing from this list - delete_on_termination, device_index, groups.

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 2]
```diff
@@ -75,65 +85,123 @@
...
     description:
       - The EC2 launch template to base instance configuration on.
-      - I(launch_template.id) the ID or the launch template (optional if name is specified).
-      - I(launch_template.name) the pretty name of the launch template (optional if id is specified).
-      - I(launch_template.version) the specific version of the launch template to use. If unspecified, the template default is chosen.
+    type: dict
+    suboptions:
+      id:
+        description:
+        - the ID or the launch template (optional if name is specified).
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py]: I think this was a typo before
```suggestion
        - the ID of the launch template (optional if name is specified).
```

File: test/integration/targets/ec2_instance_a/tasks/external_resource_attach.yml [hunk 3]
```diff
@@ -115,27 +99,32 @@
...
     - ec2_key:
         name: "{{ resource_prefix }}_test_key"
         state: absent
-        <<: *aws_connection_info
+      ignore_errors: yes
+
+    - ec2_eni:
+        eni_id: '{{ eni_a.interface.id }}'
+        state: absent
+      ignore_errors: yes
```
Comment (s-hertel) [inline test/integration/targets/ec2_instance_a/tasks/external_resource_attach.yml]: You could put these in a loop. Is delete_on_termination not working for the ec2_eni module? According to the documentation that may only work upon modification of ENIs.

File: test/integration/targets/ec2_instance_a/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+---
+# defaults file for ec2_instance
+ec2_instance_name: '{{ resource_prefix }}-node'
+ec2_instance_owner: 'integration-run-{{ resource_prefix }}'
+ec2_instance_type: 't3.micro'
+ec2_ami_name: 'amzn2-ami-hvm-2.*-x86_64-gp2'
+
+vpc_name: '{{ resource_prefix }}-vpc-a'
+vpc_seed: '{{ resource_prefix }} a'
+vpc_cidr: '10.{{ 256 | random(seed=vpc_seed) }}.0.0/16'
```
Comment (s-hertel) [inline test/integration/targets/ec2_instance_a/defaults/main.yml]: Nice.

File: test/integration/targets/ec2_instance_a/tasks/main.yml [hunk 5]
```diff
@@ -1,12 +1,41 @@
...
-- block:
+- name: Wrap up all tests and setup AWS credentials
+  module_defaults:
+    group/aws:
+      aws_access_key: "{{ aws_access_key }}"
+      aws_secret_key: "{{ aws_secret_key }}"
+      security_token: "{{ security_token | default(omit) }}"
+      region: "{{ aws_region }}"
+    ec2_instance:
+      wait: yes
```
Comment (s-hertel) [inline test/integration/targets/ec2_instance_a/tasks/main.yml]: I think this is the default already.

**Review by s-hertel — COMMENTED** [review]
This is looking good. Thanks for cleaning up so much.

File: test/integration/targets/ec2_instance_a/tasks/external_resource_attach.yml (hunk 3 above)
Comment (tremble) [reply test/integration/targets/ec2_instance_a/tasks/external_resource_attach.yml]: They didn't go away on termination when I was testing...

File: test/integration/targets/ec2_instance_a/tasks/main.yml (hunk 5 above)
Comment (tremble) [reply test/integration/targets/ec2_instance_a/tasks/main.yml]: So it is...

File: test/integration/targets/ec2_instance/tasks/iam_instance_role.yml [hunk 6]
```diff
@@ -77,26 +66,34 @@
...
         security_groups: "{{ sg.group_id }}"
-        instance_type: t2.micro
+        instance_type: "{{ ec2_instance_type }}"
         instance_role: "{{ iam_role_2.arn.replace(':role/', ':instance-profile/') }}"
-        <<: *aws_connection_info
       register: instance_with_updated_role
-      until: instance_with_updated_role is not failed
-      retries: 10
+
+    # XXX We shouldn't need this
```
Comment (s-hertel) [inline test/integration/targets/ec2_instance/tasks/iam_instance_role.yml:75]: The associated instance role not updating immediately isn't an error I've seen crop up, but I assume you have?

**Review by s-hertel — APPROVED** [review]

File: test/integration/targets/ec2_instance/tasks/iam_instance_role.yml (hunk 6 above)
Comment (tremble) [reply test/integration/targets/ec2_instance/tasks/iam_instance_role.yml:75]: It popped up once in the 10 or so runs I tried

**Comment by s-hertel** [comment]
Thanks @tremble!

**Pull request merged**